### ISSUE TITLE
Redshift timestamp column is not supported

### ISSUE THREAD

**Issue by johnchenghk01** [awslabs/aws-athena-query-federation#72]
Although document said it is supported, it is not for Redshift.

https://github.com/awslabs/aws-athena-query-federation/tree/master/athena-jdbc#data-types-support
```
NOT_SUPPORTED: Unsupported Arrow Type [Timestamp(MILLISECOND, null)] in Lambda Data Source
```

**Comment by avirtuos** [comment]
@abhijeetvg can you take a look at this when you have a moment?

**Comment by sdorazio** [comment]
We are experiencing the same exception with Aurora (MySQL-5.6) while accessing any table with a `datetime` column, even if the column is not included in the select statement. Tables without a `datetime` column work as expected.

The lambda itself does execute successfully based on the CloudWatch Metrics & Logs. However, it appears Athena does not handle the response properly.

**Comment by abhijeetvg** [comment]
We apologize for the inconvenience, this is a documentation issue. Athena does not support Arrow `timestamp` datatype to which JDBC `timestamp` should map to. There could be some remnant mappings to `DateMilli` which are incorrect. As that may result in loss of granularity supported beyond milliseconds.

This feature will be available in future, we don't have an estimated completion date yet.

**Comment by sdorazio** [comment]
I do not agree with labeling this strictly as a documentation issue. I understand the lack of support based on the timestamp type. However, I do not agree with the error handling on the Athena side.

If a table has a timestamp column, the table's metadata causes Athena to throw an exception even if the query is not using that column in any way. While I understand the intent of this, I do not agree with the behavior. This leaves a large number of tables unusable, as many systems add date created/updated timestamps for record history.

There are a few possible solutions / workaround that I can see:

1. Update Athena to not throw the exception while analyzing the table metadata. This seems like the temporary solution, but a reasonable one. Once timestamp support is added, the exception can be removed entirely.
2. Update the lambda code to not return columns of timestamp type when describing table metadata. This seems like a complete hack, but it should allow the tables to be queried as long as those columns are not used.

**Comment by avirtuos** [comment]
I tend to agree with #2 as a good interim option while we add support for more datatypes on the Athena side.

**Comment by Ceph97** [comment]
> I do not agree with labeling this strictly as a documentation issue. I understand the lack of support based on the timestamp type. However, I do not agree with the error handling on the Athena side.
> 
> If a table has a timestamp column, the table's metadata causes Athena to throw an exception even if the query is not using that column in any way. While I understand the intent of this, I do not agree with the behavior. This leaves a large number of tables unusable, as many systems add date created/updated timestamps for record history.
> 
> There are a few possible solutions / workaround that I can see:
> 
>     1. Update Athena to not throw the exception while analyzing the table metadata. This seems like the temporary solution, but a reasonable one. Once timestamp support is added, the exception can be removed entirely.
> 
>     2. Update the lambda code to not return columns of timestamp type when describing table metadata. This seems like a complete hack, but it should allow the tables to be queried as long as those columns are not used.

Do you have like a sample LAMBDA function to ignore fields with timestamp type

**Comment by avirtuos** [comment]
The issue has both documentation and bug labels. There is a proposed workaround that we are yet to implement described above to essentially ignore unsupported column types.

**Comment by avirtuos** [comment]
@atennak1 @sethtdenney think we can prioritize ignoring unsupported types so we don't throw? I'm thinking this is best done Lambda side though more annoying since it would be per-connector than doing it Athena side.

**Comment by sethtdenney** [comment]
My view is that pushing down this logic to each connector is the preferable option, but I'd expand it to add an environment variable to allow Lambda owners to explicitly signal which behavior they'd prefer. In general, I'm against implicit behavior which silently ignores data, errors, outliers, etc. as it can lead to confusion and operational pain for our customers. If we were to add in this behavior on the Athena side, then any customers whose data is unsupported but who isn't aware of that could easily run into such confusion and lack an obvious way forward when some columns are silently hidden.

@sdorazio's point that these timestamps are very common, precluding many users' tables from being at all usable, does seem to warrant prioritizing this change, so we will treat it as such.

**Comment by avirtuos** [comment]
@sethtdenney I agree that silent behaviors can lead to confusion. A happy middle ground is probably to treat unsupported columns as 'strings' and if they are queried return a value of 'NOT_SUPPORTED' this accomplishes two things: 1) no failures, 2) very obvious what is going on.

Thoughts?

**Comment by avirtuos** [comment]
I chatted with @sethtdenney offline and will be raising a PR momentarily with a change that will:
1. Add explicit validation to all connectors to ensure we don't accidentally return unsupported Types. This should surface this issue earlier in connector development for future mismatches.
2. If an unsupported type is encountered for the JDBC connector it will be ignored and we will log an error.
3. It also seems the docdb connector is vulnerable to the same issue so I'll include fixing that in this PR as well.

**Comment by fysp11** [comment]
Is there any way I can write something to my connector to allow handling these formats?
